File: Python/German_credit.py
# -*- coding: utf-8 -*-
"""
Spyder Editor

This is a temporary script file.
"""

import numpy as np
import pandas as pd
import matplotlib as plt
import re

text_file = open("/Users/chris_heale/Desktop/Data_science/Data/German_Credit/german.doc", "r")
German_schema = text_file.readlines()
text_file.close()
#Take the lines in german_schema when the regexp search is matched and place it in a list"
Schema=[line for line in German_schema if re.search("A[0-9][0-9]*",line) ]
#Pull out just the number part from each line"
LHS=[re.findall("A[0-9][0-9]*",line)[0] for line in Schema]
#Pull out the explaination part to the right of the :"
RHS=[re.split(": ",line)[1] for line in Schema]
#remove the newline on the end
RHS=[re.sub("\n","",line) for line in RHS]
#Find the column headers (we know that they are the line after Atributte then strip whitespace
headercol=[German_schema[i+1] for i in range(len(German_schema)) if re.search(r"Attribute [0-9]",German_schema[i])]
headercol=[re.sub("\n","",line) for line in headercol]
headercol=[s.lstrip() for s in headercol]

#load the German credit data as a dataframe with header cols as headers (header=None otherwise)
German_credit=pd.read_table("/Users/chris_heale/Desktop/Data_science/Data/German_Credit/germancredit.data",sep=' ',names=headercol,index_col=False)

#use dictionary to re-assign values of A** using LHS and RHS
dicti = dict(zip(LHS, RHS))
for i in range(len(headercol)): 
    if type(German_credit[headercol[i]][0])==str:
      German_credit[headercol[i] ].replace(dicti, inplace=True)
  

File: R/german_credit.r
# read in data line by line from a text file
Germ_schema<-readLines('/Users/chris_heale/Desktop/Data_science/Data/German_Credit/german.doc')
#Use reg expressions to find all the lines with A** in it.
schema<-grep('A[0-9]',Germ_schema,value = T)
#Use regular expressions to find all the A values and store note that \\1 finds the first grouping (in brackets)
LHS<-sub(".*(A[0-9][0-9].) :(.*)","\\1",schema)
#Use regular expressions to find all the meanings and store, note tat \\2 finds the second grouping (in brackets)
LHS<-sub(".*(A[0-9][0-9]*) :(.*)","\\2",schema[1:3])
#Alternately you can use for the A**
LHS<-regmatches(schema,regexpr("A[0-9][0-9]*",schema))
#And for the meanings
RHS<-regmatches(schema,regexpr(":.*",schema))
#Find the headers and add the classification variable
headers<-str_trim(Germ_schema[grep("Attribute [0-9]",Germ_schema)+1],side='left')
headers<-c(headers,"Good Loan")

#Load the german credit data itself
German_credit<-read.table("/Users/chris_heale/Desktop/Data_science/Data/German_Credit/germancredit.data",stringsAsFactors=F,header=F,sep='')
#Assign column names
colnames(German_credit)<-headers
#Assign A** values to the German Credit
for (i in 1:dim(German_credit)[1]){
  for (j in 1:dim(German_credit)[2]){
    if (!is.na(match(German_credit[i,j],LHS))){
    German_credit[i,j]<-RHS[match(German_credit[i,j],LHS)]}}}


